[MRG+1] Remove heavy memory footprint in BaseBagging due to OOB scoring

staubda opened the pull request:
Fixes #6877.
- Remove `estimators_samples` attribute from `BaseBagging`. Replace with
  `estimators_samples_` method wrapped with `@property` decorator. The 
  new method calls `_get_estimators_data_draws` described below.
- Add '_seeds' attribute to `BaseBagging` to store seeds each estimator 
  is initialized with.
- Add method `_get_estimators_data_draws` to `BaseBagging` that
  returns a generator producing sample and feature indices (or masks if desired) 
  on demand.
- Slight refactor of `_parallel_build_estimators()` to
  isolate and group lines accessing random state.

Commit:
Remove heavy memory footprint for OOB scoring from bagging.

- Remove `estimators_samples` attribute from `BaseBagging`
- Add method `_get_estimators_samples` to `BaseBagging` that
  returns a generator producing sample indices on demand.
- Slight refactor of `_parallel_build_estimators()` to
  isolate and group lines accessing random state.

Commit:
Replaced `BaseBagging.estimators_samples_` attribute with property.

- Accessing `BaseBagging.estimators_samples_` now triggers call to
  `BaseBagging._get_estimators_samples` as well as deprecation
  warning for release 0.20.
- This should make the fix for the OOB memory issue fully backwards
  compatible.

Commit:
Remove print statement.

Commit:
Added test, made new method more general.
- Added test to `test_bagging.py` to ensure that indices generated
  on the fly are identical to indices generated at fit time.
- Refactored `_get_estimators_samples()` to `_get_estimators_indices()`,
  now both feature and samples indices are returned.
- Refactored `estimators_samples_()` to deal with above.

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,6 +33,33 @@
 MAX_INT = np.iinfo(np.int32).max
 
 
+def _draw_indices(random_state, bootstrap, n_pop, n_samp):
```
You should best copy the nomenclature from `forest.py`

staubda replied:
ok so instead of 
`def _draw_indices(random_state, bootstrap, n_pop, n_samp):`
use
`def _draw_indices(random_state, bootstrap, n_population, n_samples):`
?

jnothman replied:
And probably "generate" rather than "draw", unless there's a substantial difference in meaning?

staubda replied:
Sure, no problem.

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -69,18 +95,15 @@ def _parallel_build_estimators(n_estimators, ensemble, X, y, sample_weight,
         seed = random_state.randint(MAX_INT)
```
I'm only skimming, but shouldn't these be coming from `self._seeds`?

staubda replied:
Possibly they could, but I decided to leave it like it was so that I was changing as few lines as possible. Also I'd then need to pass the `starts` variable to `_parallel_build_estimators()` when its called within `BaseBagging._fit()`.

The result should be the same either way.

jnothman replied:
Sorry, I didn't correctly see where this was coming from. Looks good.

staubda committed:
Rename index generating functions and arguments.

staubda commented:
Ok cool. My only concern now is that if someone tries to access `BaseBagging.estimators_samples_` using an index, i.e.
```python
bagging = BaseBagging().fit(X, y)
bagging.estimators_samples_[5]
```
the code will break () since `estimators_samples_` is now a generator, whereas before that would have worked.

This wouldn't be a problem for, e.g., `BaggingClassifier` which accesses `estimators_samples_` like
``` python
for samples in self.estimators_samples_:
   do      
   stuff...
```
but it might not play nicely with some user code.

jnothman commented:
If `estimators_samples_` is being deprecated, then it should *not* be accessed by `BaggingClassifier` or other internal code. It should also return exactly what it returned before.

staubda commented:
Ok, in that case I'll need to revert `estimators_samples_` to how it was before. I can't see a way to make the generator based solution exactly mimic what was there before.

Unfortunate since that will keep the unnecessary memory usage of `BaseBagging`.

jnothman commented:
Huh? I don't get it... Hop onto gitter and we'll try get this straight.

staubda commented:
just sent you a message on gitter

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +397,44 @@ def _validate_y(self, y):
+            yield feature_indices, sample_indices
+
+    @property
+    @deprecated("Attribute 'estimator_samples_' is deprecated and will be removed "
```
Should be `estimators_samples_`

staubda committed:
`estimators_samples_()` returns list instead of generator

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +397,58 @@ def _validate_y(self, y):
+
+    @property
+    @deprecated("Attribute 'estimator_samples_' is deprecated and will be removed "
+                "in release 0.20. Use method '_get_estimators_data_draw()' instead.")
```
We can't advise to use a private method. If we need to continue to provide users with access, we can not deprecate this (just regenerate it at every access), or we can provide a generator as a new attribute/method.

staubda replied:
Makes sense, as long as you're ok with it I'll just remove the deprecation warning.

Commit:
Removed `estimators_samples_` deprection warning.

Commit:
Actually removed `estimators_samples_` deprecation warning.

amueller commented on sklearn/ensemble/bagging.py:
```diff
@@ -443,8 +443,6 @@ def _get_estimators_data_draws(self, sample_mask=False, feature_mask=False):
             yield feature_draw, sample_draw
 
     @property
-    @deprecated("Attribute 'estimator_samples_' is deprecated and will be removed "
```
I think for trees we just removed and didn't replace by anything.
If we think this is important, we could deprecate this and make it a method, to make it clear that it is generated at each access. Though I'm not sure if the deprecation is worth the sugar.

amueller commented:
maybe @glouppe or @arjoly want's to have a look?

staubda commented:
If we add a new public method to `BaseBagging` does that mean we have to add to the docstrings for all classes that inherit from `BaseBagging`?

amueller commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +397,56 @@ def _validate_y(self, y):
+                                                                  max_samples)
+
+            # Convert feature indices to mask if desired
+            if feature_mask:
```
don't we have a helper function for this? I guess we have ``save_mask`` which does the opposite. this seems like such a common pattern..

staubda replied:
I'm happy to use any existing functionality you can point me to (I'm not really that familiar with the codebase), but in the original code it was being done manually so I just copied that.

amueller commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +397,56 @@ def _validate_y(self, y):
+        for seed in self._seeds:
+            # Operations accessing random_state must be performed identically
+            # to those in `_parallel_build_estimators()`
+            random_state = check_random_state(seed)
```
I'd say it's slightly more idiomatic to use ``np.random.RandomState()`` here as we know seed is an integer.

staubda replied:
So instead do 
`random_state =np.random.RandomState(seed)` ?

I just want to make sure that everything involving `random_state` is performed identically to how it's accessed in `_parallel_build_estimators`.

amueller commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +397,56 @@ def _validate_y(self, y):
 
+    def _get_estimators_data_draws(self, sample_mask=False, feature_mask=False):
+        # If max_samples is float
+        if not isinstance(self.max_samples, (numbers.Integral, np.integer)):
```
why do we need to do input validation here again? This is an internal method? I think we should store ``max_sample`` and ``max_features`` in ``self._max_samples`` and ``self._max_features`` during fit and use these here.

staubda replied:
Definitely agree with this one.

amueller commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -663,3 +663,21 @@ def test_oob_score_removed_on_warm_start():
     assert_raises(AttributeError, getattr, clf, "oob_score_")
+
+
+def test_consistent_index_sampling():
```
maybe also check that when fixing the random state, the results are identical? (we should really have that as a common test though)

staubda replied:
Do you mean make a new test where we train two separate estimators of the same type on identical data with identical random state and then see if prediction results, or oob scores are the same?

amueller replied:
well or the same one twice.

staubda replied:
Oh yeah, that would be easier.

Commit:
New mask generation function, new bagging test, new BaseBagging attributes.
- Added new private function in bagging.py that converts indices to
  a boolean mask.
- Added new bagging test to make sure identical OOB scores are generated
  when the same estimator if fit with fixed random state and identical
  training data.
- Added new private attributes `BaseBagging._max_features` and
  `BaseBagging._max_samples` to store validated input values.

staubda commented:
Questions for my latest commit:

1. Should my new indices-to-mask conversion function `_generate_mask_from_indices()` be a public function in some other module? Does it already exist in some other module?
2. Can I get rid of the validation code for `max_samples` and `max_features` in `_parallel_build_estimators` by assuming that I'll have access to my new `_max_samples` and `_max_features` attributes from `BaseBagging`?

staubda commented:
Any thoughts on the above? To deprecate or not to deprecate?

jnothman commented:
1. I have not used bagging much myself. How useful do you believe it is for the user to have access to the sample indices? If you can imagine it is necessary for some model inspection / error analysis task, then we should make it available. If we're currently making it available, I don't see the harm in leaving it available, but constructed dynamically to avoid memory usage. I would even leave it as a property rather than a method, but maybe add a note that it can be slow to fetch.
2. `_parallel_build_estimators` is private, so as long as its functionality is maintained, it's fine.
3. I'll have a look when I've got an internet connection...
4. Class docstrings don't describe methods

Commit:
Streamlined code, improved `estimators_samples_()` documentation.
- Removed `max_samples` argument from `_parallel_build_estimators()`,
  this value is now accessed via `ensemble._max_samples`.
- Removed validation of `ensemble.max_features` and `max_samples`,
  instead use `ensemble._max_features` and `ensemble._max_samples`
  which are assumed to be already validated.
- Removed unnecessary `samples` variable from `_parallel_build_estimators()`.
- Changed the way `random_state` is generated in `_parallel_build_estimators()`
  and `BaseBagging._get_estimators_data_draws()` to direct numpy method to
  reflect that seeds created in `BaseBagging._fit()` are trustworthy.
- Due to above removed generation of new seed for each estimator in
  `_parallel_build_estimators()` and `BaseBagging._get_estimators_data_draws()`.
- Added documentation to `BaseBagging.estimators_samples_()` property
  indicating the reason it's generated dynamically and the associated performance
  penalty.
- Returned `BaggingClassifier._set_oob_score()` and `BaggingRegressor._set_oob_score()`
  to directly accessing `self.estimators_samples_`.

Commit:
Streamlined code, improved `estimators_samples_()` documentation.
- Removed `max_samples` argument from `_parallel_build_estimators()`,
  this value is now accessed via `ensemble._max_samples`.
- Removed validation of `ensemble.max_features` and `max_samples`,
  instead use `ensemble._max_features` and `ensemble._max_samples`
  which are assumed to be already validated.
- Removed unnecessary `samples` variable from `_parallel_build_estimators()`.
- Changed the way `random_state` is generated in `_parallel_build_estimators()`
  and `BaseBagging._get_estimators_data_draws()` to direct numpy method to
  reflect that seeds created in `BaseBagging._fit()` are trustworthy.
- Due to above removed generation of new seed for each estimator in
  `_parallel_build_estimators()` and `BaseBagging._get_estimators_data_draws()`.
- Added documentation to `BaseBagging.estimators_samples_()` property
  indicating the reason it's generated dynamically and the associated performance
  penalty.
- Returned `BaggingClassifier._set_oob_score()` and `BaggingRegressor._set_oob_score()`
  to directly accessing `self.estimators_samples_`.

staubda commented:
>I have not used bagging much myself. How useful do you believe it is for the user to have access to the sample indices?

I haven't really used it either, I'll just leave it as a property unless somebody else comes along and has an opinion otherwise. Sounded like @amueller was ok with the property approach.

My only remaining question is whether or not `_generate_mask_from_indices()` should be a public function in some different utils type module, or if there's an existing function that provides the same functionality.

Overall I think this is good to go, gonna change to MRG, could use a thorough review though.

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -17,7 +17,7 @@
 from ..metrics import r2_score, accuracy_score
 from ..tree import DecisionTreeClassifier, DecisionTreeRegressor
-from ..utils import check_random_state, check_X_y, check_array, column_or_1d
+from ..utils import check_random_state, check_X_y, check_array, column_or_1d, deprecated
```
pep8

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +33,55 @@
+    return indices
+
+
+def _generate_bagging_indices(random_state, bootstrap_features, bootstrap_samples, 
```
pep8 line length

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +33,55 @@
+
+
+def _generate_bagging_indices(random_state, bootstrap_features, bootstrap_samples, 
+                              n_features, n_samples, max_features, max_samples):
```
"

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +33,55 @@
 
 
+def _generate_mask_from_indices(indices, mask_length, inverse=False):
+    """Private function used to convert list of indices to boolean mask"""
```
Don't need "private function used to"

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +33,55 @@
+
+
+def _generate_indices(random_state, bootstrap, n_population, n_samples):
+    """Private function used to draw randomly sampled indices."""
```
Don't need "private function used to"

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +33,55 @@
+
+def _generate_bagging_indices(random_state, bootstrap_features, bootstrap_samples, 
+                              n_features, n_samples, max_features, max_samples):
+    """Private function used to randomly draw feature and sample indices."""
```
Don't need "private function used to"

staubda replied:
I wrote it this way to be consistent with `_parallel_build_estimators`, `_parallel_predict_proba`, etc. 

Should I edit the docstrings for those functions too?

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +33,55 @@
 MAX_INT = np.iinfo(np.int32).max
 
 
+def _generate_mask_from_indices(indices, mask_length, inverse=False):
```
I'd be happy with `_indices_to_mask` for a utility function. I'm okay to see this as `indices_to_mask` in `utils` somewhere. I also don't really think `inverse` provides a lot, when I'd even prefer the implementation to just involve a `logical_not` or a `~` at the end.

staubda replied:
Ok, I'll get rid of the `inverse` argument. Any idea where in `utils` this would go?

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +404,38 @@ def _validate_y(self, y):
+
+    @property
+    def estimators_samples_(self):
+        """Return a dynamically generated list of boolean masks identifying 
```
Cut the initial description back to one line, as in PEP257, then leave a blank line before a fuller description. Can copy the initial description from the Attributes section of the class docstring.

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +404,38 @@ def _validate_y(self, y):
+            # Operations accessing random_state must be performed identically
+            # to those in `_parallel_build_estimators()`
+            random_state = np.random.RandomState(seed)
+            feature_indices, sample_indices = _generate_bagging_indices(random_state, 
```
line length

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +404,38 @@ def _validate_y(self, y):
+            # to those in `_parallel_build_estimators()`
+            random_state = np.random.RandomState(seed)
+            feature_indices, sample_indices = _generate_bagging_indices(random_state, 
+                                                                  self.bootstrap_features,
```
pep8 indentation and line length

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +404,38 @@ def _validate_y(self, y):
+        Note: the list is re-created at each call to the property in order to reduce 
+        the object memory footprint by not having it store the sampling data. Thus 
+        fetching the property may be slower than expected."""
+        return [sample_mask for _, sample_mask in 
```
The annoyance in terms of memory usage is that this could be a sparse matrix rather than a dense array of sample masks...

jnothman replied:
Maybe that's enough reason for deprecating it and replacing it with an attribute with sparse output...

staubda replied:
For me the main issue is that whether it's a dense matrix or sparse I don't want to store it on the object. That will cause the memory footprint of the object to be proportional to `len(X) x n_ensemble` (if I'm storing it as masks) or `n_samples x n_ensemble` (if I'm storing as indices). 

Either way that can get big pretty fast.

jnothman commented on sklearn/ensemble/bagging.py:
```diff
@@ -387,6 +404,38 @@ def _validate_y(self, y):
         # Default implementation
         return column_or_1d(y, warn=True)
 
+    def _get_estimators_data_draws(self, sample_mask=False, feature_mask=False):
```
I'm not convinced that you need those parameters for a private method used once in the code and once in testing.

staubda replied:
I can remove them

jnothman commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -663,3 +663,28 @@ def test_oob_score_removed_on_warm_start():
+    bagging.fit(X, y)
+
+    feat_inds_list = bagging.estimators_features_
+    feat_samp_inds_gen = bagging._get_estimators_data_draws()
```
why not just use `estimators_samples_`?

staubda replied:
can't use `estimators_samples_` because its not stored at fit time

jnothman replied:
huh? but shouldn't it be available any time after `fit()`?

staubda replied:
It can be dynamically generated any time after `fit()`, but I want to make sure the exact values that were generated during `fit()` (which, for `estimators_features_`, are still stored on the object) can be reproduced at a later time.

Since the whole idea of the PR is to not store `estimators_samples_` on the object any more it's not possible to use it for the consistency test.

jnothman replied:
I still don't get why you can't use `estimators_samples_` here; or rather, why it's not equivalent to testing `_get_estimators_data_draws` (ignoring low-scale memory consumption during testing) except using public API.

staubda replied:
When I call 
```python
feat_inds_list = bagging.estimators_features_
```
I'm retrieving the exact list of feature indices that were created within `_parallel_build_estimators()` during the call to `BaseBagging._fit()`.

When I call
```python
feat_samp_inds_gen = bagging._get_estimators_data_draws()
```
I'm dynamically re-generating both the sample and feature indices. I could also call `BaseBagging.estimators_samples_` as a way of dynamically generating just the sample indices, however there is no static version to compare them to.

The only statically generated set of indices left at this point is `BaseBagging.estimators_features_`. Thus I compare the values stored in that attribute against its dynamically generated counterpart, which can only be accessed by calling `BaseBagging._get_estimators_data_draws()`.

jnothman replied:
sorry, I see now. good

staubda commented:
Thanks for the review @jnothman, I'll push the requested changes tomorrow (it's bed time in my part of the world).

Commit:
PEP8, remove optional arguments, add public function to utils.
- Update indentation and line length to conform to PEP8.
- Update `estimators_samples_()` docstring to conform to PEP8 and
  PEP257.
- Remove optional arguments to `BaseBagging._get_estimators_data_draws()`,
  rename to `BaseBagging._get_estimators_indices()`.
- Remove optional arguments from `_generate_mask_from_indices()`,
  move to `utils.metaestimators`, rename to `indices_to_mask()`.

Commit:
Add indices_to_mask to __all__ in utils.metaestimators

jnothman commented:
I think there should be (and should have already been a test for `estimator_samples_`, at least its value if not its consistency with that used in training...

staubda commented:
Can you be more specific about what kind of test you'd like to see? I could do something similar to my oob test where I make sure the same results are produced when a model is fit twice on the same data.

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -70,3 +71,14 @@ def if_delegate_has_method(delegate):
     return lambda fn: _IffHasAttrDescriptor(fn, '%s.%s' % (delegate, fn.__name__))
+
+
+def indices_to_mask(indices, mask_length):
```
Thinking about it, this is more relevant to validation (i.e. different input scenarios yield different data formats) than to metaestimators

jnothman commented:
> Can you be more specific about what kind of test you'd like to see? I could do something similar to my oob test where I make sure the same results are produced when a model is fit twice on the same data.

Is there a way to check that the format is correct, and that at least one tree, re-fit with the samples and features available in the Bagging estimator's attributes, and the correct random state, yields the same tree...?

Commit:
Move utils function, new test for `estimators_samples` correctness.
- Moved `indices_to_mask` from `utils.metaestimators` to `utils.validation`
- Added new test to `ensemble.test_bagging` to make sure refitting
  an individual estimator from the bagging ensemble using the corresponding
  samples identified in `BaseBagging.estimators_samples` returns the same model.

jnothman commented:
Ha. Now I'm not so sure about it being in `utils.validation`, but maybe it should be if it accepted either masks or index. Lots of stuff just lands up in `utils.__init__` which I think is not great, but it works. Really @amueller should have input here as it's his suggestion.

jnothman commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -663,3 +663,64 @@ def test_oob_score_removed_on_warm_start():
+    bagging.fit(X, y)
+
+    # Get attributes, convert possible generators to lists
+    estimators_samples = list(bagging.estimators_samples_)
```
It should already be a list.

jnothman commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -663,3 +663,64 @@ def test_oob_score_removed_on_warm_start():
+    # Test for correct formatting
+    assert_equal(len(estimators_samples), len(estimators))
+    assert_equal(len(estimators_samples[0]), len(X))
+    assert_true(isinstance(estimators_samples[0][0], (np.bool_, bool)))
```
should we be testing more specifically that `estimators_samples[0]` is an array of booleans?

staubda replied:
that's kind of what I was going for with the last two tests. is there an easy way to test for that more directly?

jnothman replied:
the length thing is fair enough. otherwise, `assert_equal(arr.dtype.kind, 'b')` might be easiest

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -5,6 +5,7 @@
 
 from operator import attrgetter
 from functools import update_wrapper
+import numpy as np
```
Please revert all changes to this file

jnothman commented:
With those minor things dealt with, I'm +1 for this.

staubda commented:
Ok, will do. 

Also, we could probably remove `BaseBagging._get_estimators_indices()` and just stick the necessary bits into `BaseBagging.estimators_samples()`. We'd have to get rid of the `test_consistent_index_sampling()` test, but `test_estimators_samples()` is now doing pretty much the same thing.

Just wanted to throw it out there.

jnothman commented:
I think keep the private generator, and the public attribute producing a list (I'd rather it produced a different format, but this is backwards compatibility). I'd be happy for `test_consistent_index_sampling` to disappear.

Commit:
Adjust/remove tests, move `indices_to_mask` to `utils.__init__`.
- Remove test_bagging.test_consistent_index_sampling
- Update formatting checks in test_bagging.test_estimators_samples
  to make sure each mask is numpy boolean array
- Move indices_to_mask from utils.validation to utils.__init__

staubda commented:
This is ready for another review whenever someone gets a chance.

jnothman commented:
As a minor thing, that's really not your problem, could you please modify the `estimators_samples_ : ` section in the class docstring to clarify that the arrays are boolean masks? Thanks.

Otherwise, this LGTM.

staubda committed:
update class docstring for `estimators_samples_`

- Updated class docstring of `estimators_samples_` attribute for `BaggingClassifier` and `BaggingRegressor` to indicate that samples are identified with boolean masks.

staubda commented:
Ok, should be good to go!

jnothman commented:
Needs a second review.

agramfort commented:
maybe @ngoix ?

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +34,41 @@
+
+    # Draw indices
+    feature_indices = _generate_indices(random_state, bootstrap_features, 
+                                    n_features, max_features)
```
line under-indented

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -33,21 +34,41 @@
+    if bootstrap:
+        indices = random_state.randint(0, n_population, n_samples)
+    else:
+        indices = sample_without_replacement(n_population, n_samples, 
```
trailing whitespace. run pep8

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -57,30 +78,27 @@ def _parallel_build_estimators(n_estimators, ensemble, X, y, sample_weight,
+        # Draw random feature, sample indices
+        features, indices = _generate_bagging_indices(random_state,
+                                                      bootstrap_features,
+                                                      bootstrap, n_features, 
```
idem here and line below

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -304,6 +305,9 @@ def _fit(self, X, y, max_samples, max_depth=None, sample_weight=None):
         if not (0 < max_samples <= X.shape[0]):
             raise ValueError("max_samples must be in (0, n_samples]")
 
+        self._max_samples = max_samples
```
you have to add it to the docstrings

ngoix replied:
and maybe add a test to check that `_max_samples == max_samples` when `_fit()` is not used.

staubda replied:
I don't understand how this test would work as `_max_samples` doesn't even exist unless `_fit()` is called.

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -312,6 +316,9 @@ def _fit(self, X, y, max_samples, max_depth=None, sample_weight=None):
         if not (0 < max_features <= self.n_features_):
             raise ValueError("max_features must be in (0, n_features]")
 
+        self._max_features = max_features
```
same here

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -304,6 +305,9 @@ def _fit(self, X, y, max_samples, max_depth=None, sample_weight=None):
 
+        self._max_samples = max_samples
+
+        # If max_features is float
```
why this line? (this is not true)

ngoix commented on sklearn/ensemble/bagging.py:
```diff
@@ -289,6 +289,7 @@ def _fit(self, X, y, max_samples, max_depth=None, sample_weight=None):
 
         # Remap output
         n_samples, self.n_features_ = X.shape
+        self._n_samples = n_samples
```
By the way can you please make `max_samples` default to be None in `_fit()` as described in the docstrings?
And add lines below to use `max_samples = self.max_samples` if max_samples is None ?

ngoix commented:
Have we an idea on the gain in memory vs. the increase in computational time?

jnothman commented:
Thanks for picking up some minor issues, @ngoix, but I'm not sure what you mean in terms of extra computational time: the point is that this data is not needed for prediction, but was used when fitting. The computation is pretty similar in this PR and in master, but the whole data is not materialised in-memory.

ngoix commented:
Ok sorry I though the property `estimators_samples_` was called during prediction. This PR looks good to me!

Commit:
PEP8, made `max_samples` arg optional in _fit()
- Fixed PEP8 issues.
- Gave `max_samples` argument of `BaseBagging._fit()` a default
  value of `None` as indicated in the docstring. Added code to
  check for and process `None` value.
- Added a few more comments to `_fit()`.

staubda commented:
Ok, made the requested updates (except for the test, which I'm not sure how to implement). 

Regarding benchmarking, as @jnothman mentioned this PR has no computational impact on `predict` or even on `fit` unless OOB scoring is used. Also, I did some memory related benchmarking in my initial post on the issue [here](https://github.com/scikit-learn/scikit-learn/issues/6877).

ngoix commented:
About `_max_samples`:
1) This parameter (the effective sub-sampling size) is called `max_samples_` in IsolationForest. Maybe it would be good to have the same convention? 
2) This parameter always exists since `_fit()` is always called. Sometimes (e.g. in IsolationForest) `_fit()` is called with a parameter `max_samples` different from `self.max_samples` (e.g. `max_samples = max(self.max_samples, n_samples)`). It could be good to add a test in iforest checking that `IsolationForest.max_samples_ == BaseBagging._max_samples`.
 3) It could be worth to check the relation between `_max_samples` and `max_samples` (e.g. `max_samples == _max_samples` if max_samples is an integer) in the bagging tests.

staubda commented:
Ok, I can create those tests. I thought before you were asking for a comparison in some case where `_fit()` is never called.

Regarding `_max_samples` vs `max_samples_`, my preference would be to leave it as is since it currently uses the same convention as `_max_features`. Also this PR currently makes no changes to the public interface of `BaseBagging`, `BaggingClassifier` etc. which to me is a good thing. Of course I'm happy to make the change if there's a consensus that it's the right way to go.

ngoix commented:
Yes but `_max_features` was also created by this PR right? That's true that no changes in the public interface is a plus.

staubda commented:
Ha, oh yeah, forgot that I created `_max_features`.

staubda commented:
Well, I guess it can go either way then, maybe @jnothman can weigh in with his preference.

amueller commented on sklearn/utils/__init__.py:
```diff
@@ -419,3 +419,27 @@ def tosequence(x):
         return list(x)
+
+
+def indices_to_mask(indices, mask_length):
```
I'm fine with that being here.

amueller commented:
I think leaving the new attributes private is fine. This LGTM.

jnothman commented:
@staubda please mention the enhancement in `whats_new.rst`, and we'll merge this. Thanks!@

Commit:
2 new `max_samples` consistency tests, update whats_new.rst

Commit:
Minor change

jnothman commented on doc/whats_new.rst:
```diff
@@ -230,6 +230,13 @@ Enhancements
+   - The memory footprint is greatly reduced for :class:`BaseBagging` and
+     classes that inherit from it, i.e, :class:`BaggingClassifier`,
+     :class:`BaggingRegressor`, and :class:`IsolationForest`, by dynamically
+     generating attribute `estimators_samples_` only when it is needed.
```
Please use double-backticks for tt style in RST

jnothman commented:
@staubda sorry to draw this out. I was hoping to merge, but given your new tests, I'd really like to get a CI green tick first. Now it looks like you're going to need a rebase on master, then hopefully when it gets a green tick, ping @amueller or me for merge.

staubda committed:
Update whats_new.rst

jnothman commented:
I don't think that did what you wanted.

staubda commented:
Nope definitely not.

staubda commented:
Having trouble figuring out what the conflict is.

staubda commented:
Also not sure what the hell I just did.

jnothman commented:
I'd suggest something like
```
git checkout master
git pull upstream master
git checkout bagging-oob-memory-fix
git reset --hard b0f6e5d  # revert to previous version
git rebase master
...
git push -f origin bagging-oob-memory-fix
```

staubda commented:
Thanks for augmenting my crappy git skills, this is looking promising.

jnothman commented:
Abracadabra...

jnothman commented:
Thanks!

staubda commented:
My pleasure, thanks for all the guidance.

jnothman commented:
You're welcome! Turning good contributors into better contributors makes winners all round :)

ngoix commented:
Thanks @staubda!!

amueller commented:
awesome, thanks :)
